// @celo/contractkit/lib/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CeloContract;
(function (CeloContract) {
    CeloContract["Attestations"] = "Attestations";
    CeloContract["BondedDeposits"] = "BondedDeposits";
    CeloContract["Escrow"] = "Escrow";
    CeloContract["Exchange"] = "Exchange";
    CeloContract["GasCurrencyWhitelist"] = "GasCurrencyWhitelist";
    CeloContract["GasPriceMinimum"] = "GasPriceMinimum";
    CeloContract["GoldToken"] = "GoldToken";
    CeloContract["Governance"] = "Governance";
    CeloContract["MultiSig"] = "MultiSig";
    CeloContract["Random"] = "Random";
    CeloContract["Registry"] = "Registry";
    CeloContract["Reserve"] = "Reserve";
    CeloContract["SortedOracles"] = "SortedOracles";
    CeloContract["StableToken"] = "StableToken";
    CeloContract["Validators"] = "Validators";
})(CeloContract = exports.CeloContract || (exports.CeloContract = {}));
exports.AllContracts = Object.keys(CeloContract).map(function (k) { return CeloContract[k]; });
exports.NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
