// @celo/contractkit/lib/utils/tx-result.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var debug_1 = __importDefault(require("debug"));
var future_1 = require("./future");
var debug = debug_1.default('kit:tx:result');
function toTxResult(pe) {
    return new TransactionResult(pe);
}
exports.toTxResult = toTxResult;
var TransactionResult = /** @class */ (function () {
    function TransactionResult(pe) {
        var _this = this;
        this.hashFuture = new future_1.Future();
        this.receiptFuture = new future_1.Future();
        pe.on('transactionHash', function (hash) {
            debug('hash: %s', hash);
            _this.hashFuture.resolve(hash);
        })
            .on('receipt', function (receipt) {
            debug('receipt: %O', receipt);
            _this.receiptFuture.resolve(receipt);
        })
            .on('error', (function (error, receipt) {
            if (!receipt) {
                debug('send-error: %o', error);
                _this.hashFuture.reject(error);
            }
            else {
                debug('mining-error: %o, %O', error, receipt);
            }
            _this.receiptFuture.reject(error);
        }));
    }
    TransactionResult.prototype.getHash = function () {
        return this.hashFuture.wait();
    };
    TransactionResult.prototype.waitReceipt = function () {
        return this.receiptFuture.wait();
    };
    return TransactionResult;
}());
exports.TransactionResult = TransactionResult;
