// @celo/utils/lib/src/countries.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var esData = require('@umpirsky/country-list/data/es_AR/country.json');
var country_data_1 = __importDefault(require("country-data"));
var EMPTY_COUNTRY = {
    alpha2: '',
    alpha3: '',
    countryCallingCodes: [],
    currencies: [],
    displayName: '',
    emoji: '',
    ioc: '',
    languages: [],
    name: '',
    names: {},
    status: '',
};
var Countries = /** @class */ (function () {
    function Countries(language) {
        // fallback to 'en-us'
        this.language = language ? language.toLocaleLowerCase() : 'en-us';
        this.countryMap = new Map();
        this.localizedCountries = Array();
        this.assignCountries();
    }
    Countries.prototype.getCountry = function (countryName) {
        var _this = this;
        if (!countryName) {
            return EMPTY_COUNTRY;
        }
        // also ignoring EU and FX here, only two missing
        var country = this.localizedCountries.find(function (c) {
            return c.names !== undefined &&
                c.names[_this.language] !== undefined &&
                c.names[_this.language].toLowerCase() === countryName.toLowerCase();
        });
        return country || EMPTY_COUNTRY;
    };
    Countries.prototype.getCountryByPhoneCountryCode = function (countryCode) {
        if (!countryCode) {
            return EMPTY_COUNTRY;
        }
        var country = this.localizedCountries.find(function (c) { return c.countryCallingCodes && c.countryCallingCodes.includes(countryCode); });
        return country || EMPTY_COUNTRY;
    };
    Countries.prototype.getCountryByCode = function (countryCode) {
        var country = this.countryMap.get(countryCode);
        return country || EMPTY_COUNTRY;
    };
    Countries.prototype.getFilteredCountries = function (query) {
        query = query.toLowerCase();
        // Return empty list if the query is empty or matches a country exactly
        // This is necessary to hide the autocomplete window on country select
        if (!query || !query.length) {
            return [];
        }
        var lng = this.language;
        var exactMatch = this.localizedCountries.find(function (c) {
            return c.names && c.names[lng] !== undefined && c.names[lng].toLowerCase() === query.toLowerCase();
        });
        // since we no longer have the country name as the map key, we have to
        // return empty list if the search result is an exact match to hide the autocomplete window
        if (exactMatch) {
            return [];
        }
        // ignoring countries without a provided translation, only ones are
        // EU (European Union) and FX (France, Metropolitan) which don't seem to be used?
        return this.localizedCountries
            .filter(function (c) {
            return c.names &&
                c.names[lng] !== undefined &&
                (c.names[lng].toLowerCase().startsWith(query) ||
                    c.countryCallingCodes[0].startsWith('+' + query));
        })
            .map(function (c) { return c.alpha2; });
    };
    Countries.prototype.assignCountries = function () {
        var _this = this;
        // add other languages to country data
        this.localizedCountries = country_data_1.default.callingCountries.all.map(function (country) {
            // this is assuming these two are the only cases, in i18n.ts seems like there
            // are fallback languages 'es-US' and 'es-LA' that are not covered
            var names = {
                'en-us': country.name,
                'es-ar': esData[country.alpha2],
            };
            var localizedCountry = __assign({ names: names, displayName: names[_this.language] }, country);
            // use ISO 3166-1 alpha2 code as country id
            _this.countryMap.set(country.alpha2.toUpperCase(), localizedCountry);
            return localizedCountry;
        });
    };
    return Countries;
}());
exports.Countries = Countries;
