// @celo/contractkit/lib/wrappers/BaseWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BaseWrapper = /** @class */ (function () {
    function BaseWrapper(kit, contract) {
        this.kit = kit;
        this.contract = contract;
    }
    BaseWrapper.prototype.proxySend = function (methodFn) {
        var _this = this;
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return _this.wrapSend(methodFn.apply(void 0, args));
        };
    };
    BaseWrapper.prototype.proxyCall = function (methodFn) {
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return methodFn.apply(void 0, args).call();
        };
    };
    BaseWrapper.prototype.proxyCallAndTransform = function (methodFn, post) {
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            return methodFn.apply(void 0, args).call()
                .then(post);
        };
    };
    BaseWrapper.prototype.wrapSend = function (txo) {
        var _this = this;
        return {
            send: function (options) { return _this.kit.sendTransactionObject(txo, options); },
            txo: txo,
        };
    };
    return BaseWrapper;
}());
exports.BaseWrapper = BaseWrapper;
