// @celo/contractkit/lib/wrappers/BondedDeposits.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var collections_1 = require("@celo/utils/lib/src/collections");
var bn_js_1 = __importDefault(require("bn.js"));
var web3_1 = __importDefault(require("web3"));
var BaseWrapper_1 = require("../wrappers/BaseWrapper");
var Roles;
(function (Roles) {
    Roles[Roles["validating"] = 0] = "validating";
    Roles[Roles["voting"] = 1] = "voting";
    Roles[Roles["rewards"] = 2] = "rewards";
})(Roles || (Roles = {}));
var BondedDepositsWrapper = /** @class */ (function (_super) {
    __extends(BondedDepositsWrapper, _super);
    function BondedDepositsWrapper() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BondedDepositsWrapper.prototype.getAccountWeight = function (account) {
        return __awaiter(this, void 0, void 0, function () {
            var accountWeight;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contract.methods.getAccountWeight(account).call()];
                    case 1:
                        accountWeight = _a.sent();
                        return [2 /*return*/, web3_1.default.utils.toBN(accountWeight)];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.getVotingDetails = function (accountOrVoterAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var accountAddress, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.contract.methods
                            .getAccountFromDelegateAndRole(accountOrVoterAddress, Roles.voting)
                            .call()];
                    case 1:
                        accountAddress = _b.sent();
                        _a = {
                            accountAddress: accountAddress,
                            voterAddress: accountOrVoterAddress
                        };
                        return [4 /*yield*/, this.getAccountWeight(accountAddress)];
                    case 2: return [2 /*return*/, (_a.weight = _b.sent(),
                            _a)];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.getBondedDepositValue = function (account, noticePeriod) {
        return __awaiter(this, void 0, void 0, function () {
            var deposit;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contract.methods.getBondedDeposit(account, noticePeriod).call()];
                    case 1:
                        deposit = _a.sent();
                        return [2 /*return*/, this.getValueFromDeposit(deposit)];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.getBondedDeposits = function (account) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.zipAccountTimesAndValuesToDeposits(account, this.contract.methods.getNoticePeriods, this.getBondedDepositValue.bind(this))];
            });
        });
    };
    BondedDepositsWrapper.prototype.getNotifiedDepositValue = function (account, availTime) {
        return __awaiter(this, void 0, void 0, function () {
            var deposit;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.contract.methods.getNotifiedDeposit(account, availTime).call()];
                    case 1:
                        deposit = _a.sent();
                        return [2 /*return*/, this.getValueFromDeposit(deposit)];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.getNotifiedDeposits = function (account) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.zipAccountTimesAndValuesToDeposits(account, this.contract.methods.getAvailabilityTimes, this.getNotifiedDepositValue.bind(this))];
            });
        });
    };
    BondedDepositsWrapper.prototype.getDeposits = function (account) {
        return __awaiter(this, void 0, void 0, function () {
            var bonded, notified, weight, gold;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getBondedDeposits(account)];
                    case 1:
                        bonded = _a.sent();
                        return [4 /*yield*/, this.getNotifiedDeposits(account)];
                    case 2:
                        notified = _a.sent();
                        return [4 /*yield*/, this.getAccountWeight(account)];
                    case 3:
                        weight = _a.sent();
                        gold = new bn_js_1.default(0);
                        bonded.forEach(function (bond) { return (gold = gold.add(bond.value)); });
                        notified.forEach(function (bond) { return (gold = gold.add(bond.value)); });
                        return [2 /*return*/, {
                                bonded: bonded,
                                notified: notified,
                                total: { weight: weight, gold: gold },
                            }];
                }
            });
        });
    };
    // FIXME this.contract.methods.delegateRewards does not exist
    BondedDepositsWrapper.prototype.delegateRewardsTx = function (account, delegate) {
        return __awaiter(this, void 0, void 0, function () {
            var sig;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getParsedSignatureOfAddress(account, delegate)];
                    case 1:
                        sig = _a.sent();
                        return [2 /*return*/, this.contract.methods.delegateRole(Roles.rewards, delegate, sig.v, sig.r, sig.s)];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.getValueFromDeposit = function (deposit) {
        return web3_1.default.utils.toBN(deposit[0]);
    };
    BondedDepositsWrapper.prototype.getParsedSignatureOfAddress = function (address, signer) {
        return __awaiter(this, void 0, void 0, function () {
            var hash, signature;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        hash = web3_1.default.utils.soliditySha3({ type: 'address', value: address });
                        return [4 /*yield*/, this.kit.web3.eth.sign(hash, signer)];
                    case 1:
                        signature = (_a.sent()).slice(2);
                        return [2 /*return*/, {
                                r: "0x" + signature.slice(0, 64),
                                s: "0x" + signature.slice(64, 128),
                                v: web3_1.default.utils.hexToNumber(signature.slice(128, 130)) + 27,
                            }];
                }
            });
        });
    };
    BondedDepositsWrapper.prototype.zipAccountTimesAndValuesToDeposits = function (account, timesFunc, valueFunc) {
        return __awaiter(this, void 0, void 0, function () {
            var accountTimes, accountValues;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, timesFunc(account).call()];
                    case 1:
                        accountTimes = _a.sent();
                        return [4 /*yield*/, Promise.all(accountTimes.map(function (time) { return valueFunc(account, time); }))];
                    case 2:
                        accountValues = _a.sent();
                        return [2 /*return*/, collections_1.zip(
                            // tslint:disable-next-line: no-object-literal-type-assertion
                            function (time, value) { return ({ time: time, value: value }); }, accountTimes.map(function (time) { return web3_1.default.utils.toBN(time); }), accountValues)];
                }
            });
        });
    };
    return BondedDepositsWrapper;
}(BaseWrapper_1.BaseWrapper));
exports.BondedDepositsWrapper = BondedDepositsWrapper;
